src/git.rs

```diff
@@ -29,6 +29,42 @@ pub static DEFAULT_TYPES: Lazy<HashMap<&str, &str>> = Lazy::new(|| {
     m
 });
 
+// Singleton pattern
+pub fn get_repository(repo: &Path) -> Result<&Mutex<Repository>, Error> {
+    static REPO: OnceCell<Mutex<Repository>> = OnceCell::new();
```

Comment:
Any specific reason why this has to be a `Mutex`? Wouldn't a `once_cell::Lazy` be enough to prevent multiple initializations?

The mutex should only be needed in a multi-threaded environment 🤔

Reply:
I mean I'd keep it for now, whatever. It's not that the performance of those 0.1ms matter :) 
I am re-writing some stuff anyway to allow for #12. I can clean stuff like this up then.

Reply:
No reason, just not sure how Lazy is different from this. I'll check and replace if it is doable, I don't mind.

I still don't get the reasoning for sync vs unsync. I know it is required by compiler as his is static, but why? Sync shouldn't be required in a single threaded app, and I don't see multithreading here.

Reply:
Btw, Mutex is there because a sync OnceCell didn't compile without some sync primitive. Not sure Lazy is going to be any different, as it is essentially the same as OnceCell, but I'll check.

Reply:
I actually have no clue why the sync version is needed here and not the unsync version. I never got unsync to work, so I guess I am missing some fundamental Rust knowledge here... Don't worry. We can keep it like this for now.

The rework does introduce structs to handle state so this won't be needed later on anyway.

Reply:
@maksyms https://doc.rust-lang.org/reference/items/static-items.html
The first bullet point: static types has to be `Sync`. And since `Repository` doesn't implement it, you have to wrap it using `Mutex`.

Reply:
Yep, I am aware of that. What I don't understand is "why". There is no first principle that dictates that must be the case.

I've done some reading, and it seems that Rust compiler is being very defensive and requires Sync for statics even if there is no multi-threading. Of course, figuring out whether the application is multi-threaded _is_ something the compiler can do. It is more difficult to do with re-entrability - so I accept the fact that the compiler cannot figure out whether the re-entrability requirements mean that the static needs to be protected with a sync primitive. Not how I'd do it in C/C++, but at least I'm closer to understanding the reason. 👍

Reply:
I mean it would need code analysis that is beyond the borrow checker right now to determine if a function does thread out. It would have to analyze every single function body instead of just the signatures to determine the validity. You can't guarantee it if you call FFI.

So they probably just opted for making it Sync.

Reply:
@SirWindfield, threading is implemented using `libpthreads` or similar libs and supported by kernels. It is trivial for Rust compiler to figure out whether any thread creation APIs are called or not. It probably chooses not to do so, and the safe thing to do when you don't check for calls to thread-creating APIs is to require Sync on static.

Reply:
I _know_ using Mutex there is right in a Rust program. What I'm still puzzled about is _why_ - from the first principles, not just from the book. :) I think the latest parts of our discussion homed in on at least some of the reasons. :)

Reply:
Ah ok I thought you didn't know _why_ you needed the Mutex. Makes sense.
> threading is implemented using libpthreads or similar libs and supported by kernels.

I didn't know that! TIL :)